File: src/problems/021. Merge Two Sorted Lists/Solution.js
// 21. Merge Two Sorted Lists
// Accepted 76ms

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var mergeTwoLists = function(l1, l2) {
    let [p, q] = [l1, l2];
    let dummy = new ListNode(-1);
    let it = dummy;
    
    while(p && q) {
        if(p.val < q.val) {
            it.next = new ListNode(p.val);
            p = p.next;
        } else {
            it.next = new ListNode(q.val);
            q = q.next;
        }
        it = it.next;
    }
    while(p) {
        it.next = new ListNode(p.val);
        [it, p] = [it.next, p.next];
    }
    while(q) {
        it.next = new ListNode(q.val);
        [it, q] = [it.next, q.next];
    }
    
    return dummy.next;
};

File: src/problems/021. Merge Two Sorted Lists/Solution.java
// 21. Merge Two Sorted Lists
// Accepted 15ms

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode ret = new ListNode(-1);
        ListNode head = ret;
        
        while(l1 != null && l2 != null) {
            if(l1.val < l2.val) {
                ret.next = l1;
                l1 = l1.next;
            }
            else {
                ret.next = l2;
                l2 = l2.next;
            }
            
            ret = ret.next;
        }
        
        if(l1 != null)
            ret.next = l1;
        if(l2 != null)
            ret.next = l2;
        
        return head.next;
    }
}
